### PR TITLE
Implemented sorting of ReallyUserFriendlyTypes using `unidecode`

### PR DESCRIPTION
**Pull request by rohnsha0**
as discussed in https://github.com/plone/Products.CMFPlone/pull/3985#discussion_r1668309150

### CODE REVIEW

**Comment by jensens** [comment]
+ 1 for sane sorting.

But:  I vote against using the `pyICU` package, because of its lack of providing wheels together with it's  need for the C-Library. Usage on for several platforms may get a PITA. 

There is pyUCA which works in a similar way but without having this problem. What do you think? 

cc @zopyx @petschki

**Comment by zopyx** [comment]
> But: I vote against using the `pyICU` package, because of its lack of providing wheels together with it's need for the C-Library. Usage on several platforms may get a PITA.

+1

However, `pyuca` appears unmaintained, last release from 2017, future unclear. 

Proposal:
- provide a universal utility method for sorting a list of strings for a given language
- default sorting using sorted()
- if `pyuca` or `pyICU`  is installed: facilitate the sort based on the collation provide by one of those packages.

**Comment by rohnsha0** [comment]
> Proposal:
> - provide a universal utility method for sorting a list of strings for a given language
> - default sorting using sorted()
> - if pyuca or pyICU is installed: facilitate the sort based on the collation provide by one of those packages.

@zopyx then according to you, we first need to check if `pyicu` or `pyuca` is installed then sorting using their collarator else using default sorting method?

**Comment by wesleybl** [comment]
And if the [unidecode](https://github.com/avian2/unidecode) package is used:

```python
>>> from unidecode import unidecode
>>> unidecode("Águia").lower()
'aguia'
```

**Comment by rohnsha0** [comment]
> And if the [unidecode](https://github.com/avian2/unidecode) package is used:
> 
> ```python
> >>> from unidecode import unidecode
> >>> unidecode("Águia").lower()
> 'aguia'
> ```

does `unicode` provide the same func? im actually unaware of it... @wesleybl

**Comment by jensens** [comment]
> 
> does `unicode` provide the same func? im actually unaware of it... @wesleybl

Not the same, but good enough for the use-case IMO.

**Comment by jensens** [comment]
> And if the [unidecode](https://github.com/avian2/unidecode) package is used:

It is a dependency of the Plone universe anyway.

**Comment by rohnsha0** [comment]
> > And if the [unidecode](https://github.com/avian2/unidecode) package is used:
> 
> It is a dependency of the Plone universe anyway.

then performing the sorting using `unicode` is a better option... right?

**Comment by wesleybl** [comment]
> does unicode provide the same func? im actually unaware of it... @wesleybl

The `unicode` function no longer exists in Python 3. In any case, it was used for something different in Python 2.

You must use the `unidecode` function within the function passed in the `key` parameter of the sort method. For example:

```python
>>> from unidecode import unidecode
>>> list = ['Café', 'Águia', 'Água']
>>> list.sort(key=lambda string: unidecode(string).lower())
>>> list
['Água', 'Águia', 'Café']
```

**Comment by wesleybl** [comment]
> It is a dependency of the Plone universe anyway.

Anyway, I think it's worth putting it as a dependency in `setup.py`

**Comment by rohnsha0** [comment]
@wesleybl @jensens verify the changes... it worked for me!

File: plone/app/vocabularies/types.py [hunk 1]
```diff
@@ -298,9 +299,13 @@ def __call__(self, context):
             for t in ttool.listContentTypes()
             if t not in BAD_TYPES
         ]
-        items.sort()
-        items = [SimpleTerm(i[1], i[1], i[0]) for i in items]
-        return PermissiveVocabulary(items)
+        
+        print(items)
```
Comment (wesleybl) [inline plone/app/vocabularies/types.py:303]: Remove this `print`

**Review by wesleybl — CHANGES_REQUESTED** [review]
Please check for test errors.

File: plone/app/vocabularies/types.py (hunk 1 above)
Comment (rohnsha0) [reply plone/app/vocabularies/types.py]: seems like unidecode is not added to setup.py requirements

```
Missing requirements
====================
     unidecode.unidecode
```

Comment (rohnsha0) [reply plone/app/vocabularies/types.py]: the reason for other two tests being failed is unknown to me... Any help to tackle the failing test is greatly appreciated... @wesleybl

File: plone/app/vocabularies/types.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
 from zope.schema.interfaces import IVocabularyFactory
 from zope.schema.vocabulary import SimpleTerm
 from zope.schema.vocabulary import SimpleVocabulary
+from unidecode import unidecode
```
Comment (wesleybl) [inline plone/app/vocabularies/types.py:11]: Add `unidecode` as a dependency in [setup.py](https://github.com/plone/plone.app.vocabularies/blob/a7b0049757f15694fee209b2701bab08a476803a/setup.py#L45)

Run `tox -e lint` to check for lint errors.

@stevepiercy the link in

https://github.com/plone/plone.app.vocabularies/blob/master/CONTRIBUTING.rst

is broken.

**Review by wesleybl — CHANGES_REQUESTED** [review]

File: plone/app/vocabularies/types.py (hunk 2 above)
Comment (rohnsha0) [reply plone/app/vocabularies/types.py:11]: > Add unidecode as a dependency in [setup.py](https://github.com/plone/plone.app.vocabularies/blob/a7b0049757f15694fee209b2701bab08a476803a/setup.py#L45)

added

Comment (wesleybl) [reply plone/app/vocabularies/types.py:11]: In alphabetical order please.

**Comment by rohnsha0** [comment]
```
autofix pushing was disabled in the configuration. to autofix this pr, comment or apply the label "pre-commit.ci autofix" to the pull request
```

it was the error found in one of failing test

File: plone/app/vocabularies/types.py (hunk 1 above)
Comment (wesleybl) [reply plone/app/vocabularies/types.py]: > the reason for other two tests being failed is unknown to me... Any help to tackle the failing test is greatly appreciated

See: https://github.com/plone/plone.app.vocabularies/pull/93#discussion_r1688272276

File: plone/app/vocabularies/types.py (hunk 2 above)
Comment (rohnsha0) [reply plone/app/vocabularies/types.py:11]: > In alphabetical order please.

done

**Comment by rohnsha0** [comment]
pre-commit.ci autofix

**Comment by rohnsha0** [comment]
@jenkins-plone-org please run jobs

File: news/3985.chore [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Implemented pyICU for sorting of ReallyUserFriendlyTypes
```
Comment (wesleybl) [inline news/3985.chore:1]: `pyICU` not used here.

**Review by wesleybl — CHANGES_REQUESTED** [review]

File: news/3985.chore (hunk 3 above)
Comment (rohnsha0) [reply news/3985.chore]: changed

**Comment by wesleybl** [comment]
@jenkins-plone-org please run jobs

**Comment by rohnsha0** [comment]
@jenkins-plone-org please run jobs

File: news/3985.chore [hunk 4]
```diff
@@ -0,0 +1,2 @@
+Implemented sorting of ReallyUserFriendlyTypes using `unidecode`
```
Comment (wesleybl) [inline news/3985.chore:1]: Period at the end of the phase.

**Comment by rohnsha0** [comment]
@jenkins-plone-org please run jobs

**Comment by rohnsha0** [comment]
@jenkins-plone-org please run jobs

**Comment by rohnsha0** [comment]
@wesleybl all tests except one... which shows strange tracebacks

```
Setup failed:
SessionNotCreatedException: Message: session not created
from unknown error: cannot create temp dir for user data dir
```


```
Also teardown failed:
No browser is open.
```

**Review by jensens — APPROVED** [review]
+1

**Review by wesleybl — APPROVED** [review]
LGTM

**Pull request merged**